### ISSUE TITLE
R make-like and parallel behavior script

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/rstan#127]
@bgoodri suggested putting a form of Sebastian Weber's make-like script on the getting started page.  The desiderata are:

* [x] make-like behavior to save models
    - [x] for files
    - [x] for model strings

* [x] parallel execution as default

This may have to be done on a per-platform basis and perhaps not using parallel for windows.

**Comment by bgoodri** [comment]
I don't see how to do it easily for model strings. The function below expects a Stan program with a .stan extension so that it can check for the existence of a file with the same name but an .rda extension instead of a .stan extension. With just a model string, it wouldn't know what binary file to look for. I suppose we could load all files from the working directory with a .RData or .rda extension and check that get_stancode() returns a result that is identical to the passed model string, but then we could be clogging the RAM with unused objects.

In any event, testing needed ...
```
## runs Stan in parallel with caching of compiled Stan models
stan <- function(stanProgram, cores = parallel::detectCores(), pedantic = TRUE, ...) {
  if(!grepl("stan$", stanProgram)) {
    stop("'stanProgram' must be the name of a file with a '.stan' extension")
  }
  dots <- list(...)
  stanProgram.rda <- gsub("stan$", "rda", stanProgram)
  if(!file.exists(stanProgram.rda) |
      file.info(stanProgram.rda)$mtime < file.info(stanProgram)$mtime | 
      as.integer(file.info(stanProgram.rda)$mtime) <
      as.integer(as.POSIXct(packageDescription("rstan")$Date)) ) {
    if(pedantic) cat("Model needs recompilation.\n")
    stanExe <- suppressMessages(rstan::stan(stanProgram, chains = 0, data = dots$data))
    saveRDS(stanExe, file = stanProgram.rda)
  } else {
    if(pedantic) cat("Loading cached model.\n")
    stanExe <- readRDS(stanProgram.rda)
  }
  chains <- dots$chains
  if(is.null(chains)) chains <- 4L
  dots$chains <- 1
  dots$fit <- stanExe
  stopifnot(require(parallel))
  if(.Platform$OS.type != "windows") {
    out <- mclapply(1:chains, mc.cores = cores, FUN = function(i) {
      dots$chain_id <- i
      do.call(rstan::stan, args = dots)
    })
  }
  else { # Windows
    cl <- makePSOCKcluster(cores)
    on.exit(stopCluster(cl))
    clusterExport(cl, envir = environment(), varlist = c("dots"))
    clusterEvalQ(cl, expr = require(Rcpp))
    out <- parLapply(cl, X = 1:cores, fun = function(i) {
      dots$chain_id <- i
      do.call(rstan::stan, args = dots)      
    })
  }
  return(sflist2stanfit(out))
}

# Demonstration of useage:
# stopifnot(require(rstan))
# setwd(tempdir())
# stopifnot(file.copy(grep("eight_schools", stan_demo(0), value = TRUE), 
#               getwd(), overwrite = TRUE))
# schools_dat <- list(J = 8, 
#                     y = c(28,  8, -3,  7, -1,  1, 18, 12),
#                     sigma = c(15, 10, 16, 11,  9, 11, 10, 18))
# eight_schools <- stan("eight_schools.stan", data = schools_dat)
# eight_schools <- stan("eight_schools.stan", data = schools_dat, chains = 8)
# stopifnot(ncol(eight_schools) == 8)
```

**Comment by bgoodri** [comment]
Improved version, now with an option to optimize instead of sampling

```
## runs Stan in parallel (if not optimizing) with caching of compiled Stan programs
stan <- function(stanProgram, optimize = FALSE, 
                 cores = parallel::detectCores(), pedantic = TRUE, ...) {
  if(!grepl("stan$", stanProgram)) {
    stop("'stanProgram' must be a path to a file with a '.stan' extension")
  }
  dots <- list(...)
  stanProgram.rda <- gsub("stan$", "rda", stanProgram)
  if(!file.exists(stanProgram.rda) |
       file.info(stanProgram.rda)$mtime < file.info(stanProgram)$mtime | 
       as.integer(file.info(stanProgram.rda)$mtime) <
       as.integer(as.POSIXct(packageDescription("rstan")$Date)) ) {
    if(pedantic) cat("Model needs recompilation.\n")
    stanExe <- suppressMessages(rstan::stan(stanProgram, chains = 0, data = dots$data))
    saveRDS(stanExe, file = stanProgram.rda)
  } else {
    if(pedantic) cat("Loading cached model.\n")
    stanExe <- readRDS(stanProgram.rda)
  }
  if(optimize) {
    dots$object <- get_stanmodel(stanExe)
    out <- do.call(rstan::optimizing, args = dots)
    return(out)
  }
  chains <- dots$chains
  if(is.null(chains)) chains <- 4L
  dots$chains <- 1
  dots$fit <- stanExe
  callFun <- function(i) {
    dots$chain_id <- i
    return(do.call(rstan::stan, args = dots))
  }
  stopifnot(require(parallel))
  if(.Platform$OS.type != "windows") out <- mclapply(1:chains, mc.cores = cores, FUN = callFun)
  else { # Windows
    cl <- makePSOCKcluster(cores)
    on.exit(stopCluster(cl))
    clusterExport(cl, envir = environment(), varlist = "dots")
    clusterEvalQ(cl, expr = require(Rcpp))
    out <- parLapply(cl, X = 1:chains, fun = callFun)
  }
  return(sflist2stanfit(out))
}

# Demonstration of useage:
# stopifnot(require(rstan))
# setwd(tempdir())
# stopifnot(file.copy(grep("eight_schools", stan_demo(0), value = TRUE), 
#               getwd(), overwrite = TRUE))
# schools_dat <- list(J = 8, 
#                     y = c(28,  8, -3,  7, -1,  1, 18, 12),
#                     sigma = c(15, 10, 16, 11,  9, 11, 10, 18))
# eight_schools <- stan("eight_schools.stan", data = schools_dat)
# eight_schools <- stan("eight_schools.stan", data = schools_dat, chains = 8)
# stopifnot(ncol(eight_schools) == 8)
# writeLines(c('parameters { real y; }', 'model{ y ~ normal(0,1); }'), con = "normal.stan")
# normal <- stan("normal.stan", optimize = TRUE)
# normal <- stan("normal.stan", optimize = TRUE, hessian = TRUE)
```

**Comment by bgoodri** [comment]
Here is a version that works with program strings (within the same R session)

```
## runs Stan in parallel (if not optimizing) with caching of compiled Stan programs
stan <- function(stanProgram,      # path to .stan file or program string
                 optimize = FALSE, # use sampling() (default) or optimizing()
                 cores = parallel::detectCores(), # number of cores to use if sampling()
                 pedantic = TRUE,  # whether to print status messages
                 ...) {            # further arguments to sampling() or optimizing()
  if(length(stanProgram) > 1 || !grepl("stan$", stanProgram)) { # program string
    tf <- tempfile()
    writeLines(stanProgram, con = tf)
    stanProgram <- file.path(dirname(tf), paste0(tools::md5sum(tf), ".stan"))
    if(!file.exists(stanProgram)) file.rename(from = tf, to = stanProgram)
  }
  else if(!file.exists(stanProgram)) stop(paste(stanProgram, "does not exist"))

  mtime <- file.info(stanProgram)$mtime
  dots <- list(...)
  stanProgram.rda <- gsub("stan$", "rda", stanProgram)
  if(!file.exists(stanProgram.rda) |
     file.info(stanProgram.rda)$mtime <  mtime | 
     mtime < as.POSIXct(packageDescription("rstan")$Date) ) {
         
    if(pedantic) cat("Model needs compilation.\n")
    stanExe <- suppressMessages(rstan::stan(stanProgram, chains = 0, data = dots$data))
    saveRDS(stanExe, file = stanProgram.rda)
  } 
  else {
    if(pedantic) cat("Loading cached model.\n")
    stanExe <- readRDS(stanProgram.rda)
  }
  
  if(optimize) {
    dots$object <- get_stanmodel(stanExe)
    out <- do.call(rstan::optimizing, args = dots)
    return(out)
  } 
  
  # sampling()
  chains <- dots$chains
  if(is.null(chains)) chains <- 4L
  dots$chains <- 1
  dots$fit <- stanExe
  dots$refresh <- 0L
  callFun <- function(i) {
    dots$chain_id <- i
    return(do.call(rstan::stan, args = dots))
  }
  stopifnot(require(parallel))
  if(.Platform$OS.type != "windows") {
    out <- mclapply(1:chains, FUN = callFun, mc.cores = cores, mc.silent = TRUE)
  }
  else { # Windows
    cl <- makePSOCKcluster(cores)
    on.exit(stopCluster(cl))
    clusterExport(cl, envir = environment(), varlist = "dots")
    clusterEvalQ(cl, expr = require(Rcpp))
    out <- parLapply(cl, X = 1:chains, fun = callFun)
  }
  return(sflist2stanfit(out))
}

# Demonstration of useage:
# stopifnot(require(rstan))
# setwd(tempdir())
# stopifnot(file.copy(grep("eight_schools", stan_demo(0), value = TRUE), 
#               getwd(), overwrite = TRUE))
# schools_dat <- list(J = 8, 
#                     y = c(28,  8, -3,  7, -1,  1, 18, 12),
#                     sigma = c(15, 10, 16, 11,  9, 11, 10, 18))
# eight_schools <- stan("eight_schools.stan", data = schools_dat)
# eight_schools <- stan("eight_schools.stan", data = schools_dat, chains = 8)
# stopifnot(ncol(eight_schools) == 8)
# normal <- stan(c('parameters { real y; }', 'model{ y ~ normal(0,1); }'), 
#                optimize = TRUE)
# normal <- stan(c('parameters { real y; }', 'model{ y ~ normal(0,1); }'), 
#                optimize = TRUE, hessian = TRUE)

```

**Comment by bob-carpenter** [comment]
Cool!  Thanks.  I'll give these a whirl the next
time I do some modeling.

- Bob

**Comment by bgoodri** [comment]
@maverickg Along the lines of the latest version above, what do you think of changing rstan to use tools::md5sum() of the code rather than tempfile() in these two places:
```
goodrich@T540p:/opt/rstan/rstan/rstan$ git grep -n -F "tempfile" | cat
R/cxxfunplus.R:61:  tfile <- tempfile()
R/misc.R:1306:  namep1 <- if (obfuscate_name)  basename(tempfile('model', '')) else 'model'
```
That would yield a unique but deterministic name, and we could then exploit ccache and the fact that R CMD SHLIB has some make-like behavior.

**Comment by maverickg** [comment]
@bgoodri 
Just the changing the filename would not work. package inline automatically generates a function in the cpp file with name derived from tempfile().  

In fact, I tried to change that function name (see [here](https://github.com/maverickg/inline/commit/122f6daaa341693200ba19e4a60e783378ef5906)) in a modified inline package.  It works most of the time by defining CXX = ccache g++.  However, a couple of days ago, I had a segfault problem using this modified inline.  After replacing inline with the one from CRAN, the segfault problem went away.  I  didn't investigate why it happened.  Now I even don't remember where I had this segfault problem (maybe one of the tests).  But as I don't understand much about the underlying stuff about loading dynamic objects and how they get used by Rcpp, it is difficult to find out the reason.

**Comment by lionel-** [comment]
Hello,

Is it really necessary to attach the parallel package here?

``` r
stopifnot(require(parallel))
if(.Platform$OS.type != "windows") {
  out <- mclapply(1:chains, FUN = callFun, mc.cores = cores, mc.silent = TRUE)
}
```

If it's possible, it'd be nice to use `requireNamespace()` instead of `require()`, and then prepend your calls with the namespace, e.g. `parallel::mclapply()`. This way, the user search path is not modified.

**Comment by bob-carpenter** [comment]
I have no idea what "requireNamespace()" does, but if
we can bring third-party packages in without automatically
including using their namespace at the top level, 
I think that'd be great.  Especially if we don't require 
users to call those packages directly.

As to attaching rstan's namespace, I'm not sure what
the convention normally is in R.  On the one hand, I hate
all the globals R uses, but on the other hand, it seems to
be the convention.  I don't think Andrew or others are going
to be wanting to type rstan::stan(), but it'd be OK with me. 
Maybe have people bring it in explicitly?
- Bob

**Comment by lionel-** [comment]
`requireNamespace()` loads the namespace in memory if it's installed, but does not attach it to the search space.

For reference, the second paragraph in Bob's message refers to a concern I raised in #118 ;)

There would be no requirement to leave rstan unattached as it will always be possible to have `library("rstan")` in a script. The goal is to let users decide what package they want attached. Currently, it's impossible to use `stan()` without attaching rstan, which is a problem.

**Comment by bob-carpenter** [comment]
> On Mar 11, 2015, at 11:46 PM, Lionel Henry notifications@github.com wrote:
> 
> requireNamespace() loads the namespace in memory if it's installed, but does not attach it to the search space.
> 
> For reference, the second paragraph in Bob's message refers to a concern I raised in #118 ;)
> 
> There would be no requirement to leave rstan unattached as it will always be possible to have library("rstan") in a script. The goal is to let users decide what package they want attached. Currently, it's impossible to use stan() without attaching rstan, which is a problem.

That does sound like a problem if it's not even an option.
I tend not to use any other R packages so it's never been

I have to qualify all this with the fact that I'm not an
R expert and have to defer to Ben and Jiqiang to make any
decisions about this.  
- Bob

**Comment by maverickg** [comment]
I cannot discern the subtleness here.  I don't really see a big problem of loading `parallel` and when using `rstan`, it would be more convenient to attach it.

**Comment by bgoodri** [comment]
I will change the require() thing when that script gets folded into the
rest of rstan.

**Comment by lionel-** [comment]
@maverickg 
attaching the package as a side effect is not necessary and it clutters the workspace.

@bgoodri 
great!

**Comment by maverickg** [comment]
@lionel- 
thanks.  In general I don't like side effect either.  If a user has already done `library(parallel)`, would `loadNamespace(parallel)` put it in an internal database? If yes, it would use more resources.

**Comment by lionel-** [comment]
I just checked the man page, the namespace is only loaded in memory once, so no problem on this side.

> A request to load a name space when one of that name is already loaded has no effect.

**Comment by bgoodri** [comment]
This is done (albeit over various commits) for 2.7.0 but you really only get make-like behavior if you use multiple cores.
